### PR TITLE
improve inferability of base

### PR DESCRIPTION
**Pull request by vtjnash**
Avoid growing an input parameter during recursion. In general, permitting that would require that inference be potentially non-terminating (as proving that it will converge requires solving the halting problem). However, the methods in base using this pattern only actually requires the simpler pattern of tail-recursing on some element and concatenating the output.

(This is the learnings of #21933 without the policy change.)

### CODE REVIEW

**Comment by StefanKarpinski** [comment]
Is this really safe to backport to 0.6? It seems awfully large and potentially dangerous for that. It would have to be fixing some really bad issue in order to warrant that backport.

**Comment by vtjnash** [comment]
Its somewhat large, but mostly mechanical. I probably wouldn't bother if we already had a release, but I think it's useful to keep master closer during the RC period. There's a few bug fix commits in here too (1, 3, and 7) that should be backported regardless.

**Comment by vtjnash** [comment]
sysimg also seems to be about a megabyte smaller too.

**Comment by StefanKarpinski** [comment]
As always, can we have just the bug fixes in their own PRs please? Those can and should be merged as soon as they pass CI. The rest is less clear.

**Comment by vtjnash** [comment]
This is the bugfix PR (feature PR is #21933). It's just not quite as essential that we backport the correctness fixes as the assertion fixes.

**Comment by tkelman** [comment]
What bugs are being fixed? Where are the tests?

File: base/abstractarray.jl [hunk 1]
```diff
@@ -1160,30 +1158,32 @@ cat_similar(A::AbstractArray, T, shape) = similar(A, T, shape)
...
+@inline function _cshp(ndim::Int, ::Tuple{}, shape, nshape)
+    next = _cs(ndim, shape[1], nshape[1])
+    (next, _cshp(ndim + 1, (), tail(shape), tail(nshape))...)
 end
-@inline function _cshp(dims, out, shape, nshape)
-    next = _cs(length(out) + 1, dims[1], shape[1], nshape[1])
-    _cshp(tail(dims), (out..., next), tail(shape), tail(nshape))
+@inline function _cshp(ndim::Int, dims, shape, nshape)
+	a = shape[1]
+	b = nshape[1]
```
Comment (tkelman) [inline base/abstractarray.jl]: spaces, not tabs

**Comment by vtjnash** [comment]
The fixes here are to ensure that inference will always eventually terminate when any of these methods enter its queue. Tests are PR #21933 which addresses that bug more broadly.

**Comment by mbauman** [comment]
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/22019#issuecomment-303236802) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/c81ea7cd2b377f36920a5bf41e6901216c222689/c916ca9_vs_afd09f2/report.md). cc @jrevels

**Comment by tkelman** [comment]
> Tests are PR #21933 which addresses that bug more broadly.

There is exactly one line of tests changed in https://github.com/JuliaLang/julia/pull/21933/files. I'm against backporting this without any tests.

**Comment by nalimilan** [comment]
This broke `broadcast` in DataArrays, see https://github.com/JuliaStats/DataArrays.jl/pull/259#issuecomment-304627359 and https://github.com/JuliaLang/julia/pull/22130.

**Comment by MikeInnes** [comment]
cf0c6a1e5e0983f1ab11c774b54c7496d6e46650 fixes an issue for me and is trivial to backport, would like to see that in a point release.

**Comment by tkelman** [comment]
can you submit a unit test that it fixes?

**Pull request merged**